data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java

```diff
@@ -110,25 +124,37 @@ private WireEvent getWireEvent(final Event event) {
         );
     }
 
-    private CompletableFuture<AggregatedHttpResponse> processHttpRequest(final WebClient client, final String content) {
+    private CompletableFuture<AggregatedHttpResponse> processHttpRequest(final WebClient client, final String content,
+                                                                         final Collection<Record<Event>> records,
+                                                                         final PeerForwarderReceiveBuffer<Record<Event>> peerForwarderReceiveBuffer) {
         return CompletableFuture.supplyAsync(() ->
         {
-            final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
-            return aggregate.join();
-        }, executorService);
+            try {
+                final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
+                return aggregate.join();
+            } catch (final Exception e) {
+                processFailedRequestsLocally(null, records, peerForwarderReceiveBuffer);
+                throw e;
+            }
+        }, executorService).thenApply(httpResponse -> processFailedRequestsLocally(httpResponse, records, peerForwarderReceiveBuffer));
```

Comment:
I'd imagine it would be possible for the process worker to go faster than then peer requests and then these threads from the `executorService` fill up. Will the call block the processor thread when the `executorService`'s thread pool is exhausted?

Reply:
That's a good question. Let me see if I can simulate this scenario. It probably makes sense to have a dedicated threadpool here regardless

Reply:
I agree that we should have a dedicated thread-pool.

I am wanting to be sure that we handle the situation where the thread pool is exhausted correctly. If this throws an exception that would be bad. I think blocking to wait is acceptable in my opinion. And I'd be concerned about having some secret buffer of queued requests that we have no real insight into.